docs/proposals/20210524-managed-external-etcd.md

```diff
@@ -0,0 +1,570 @@
+- We can reuse this same "scale down" function when removing a healthy member during an etcd upgrade. So even upgrade will follow the same pattern of removing the member first and then adding a new one with the latest spec.
+- Once the upgrade/machine replacement is completed, the etcd controller will again update the EtcdadmCluster.Status.Endpoints field. The control plane controller will rollout new Machines that use the updated etcd endpoints for the apiserver.
+
+###### Periodic etcd member healthchecks
+- The etcdadm cluster controller will perform healthchecks on the etcd members periodically at a predetermined interval. The controller will perform client healthchecks by making HTTP Get requests to the `<etcd member address>:2379/health` endpoint of each etcd member. It cannot perform peer-to-peer healthchecks since the controller is external to the etcd cluster and there won't be any agents running on the etcd members.
+- If a member continuously fails healthcheck for a certain amount of time, the etcdadm cluster controller will replace that member with a new member.
+
+#### Changes needed in docker machine controller
+- Each machine infrastructure provider sets a providerID on every InfraMachine.
+- Infrastructure providers like CAPA and CAPV set this provider ID based on the instance metadata.
+- CAPD on the other hand first calls `kubectl patch` to add the provider ID on the node. And only then it sets the provider ID value on the DockerMachine resource. But the problem is that in case of an etcd only cluster, the machine is not registered as a Kubernetes node. Cluster provisioning does not progress until this provider ID is set on the Machine. As a solution, CAPD will check if the bootstrap provider is etcdadm, and skip the kubectl patch process and set the provider ID on DockerMachine directly. These changes are required in the [docker machine controller](https://github.com/mrajashree/cluster-api/pull/2/files#diff-1923dd8291a9406e3c144763f526bd9797a2449a030f5768178b8d06d13c795bR307) for CAPD.
+
+#### Clusterctl changes
+- External etcd topology is optional so it won't be enabled by default. Users should be able to install it through `clusterctl init` when needed. This requires the following changes in clusterctl
+- New ProviderTypes
+```go
+EtcdBootstrapProviderType = ProviderType("EtcdBootstrapProvider")
+EtcdProvider = ProviderType("EtcdProvider")
```

Comment:
This impact the operator work as well... At least let's state as a future goal the intent to update the operator proposal to accommodate those providers as well